# **PANDAS**

##DataFrame

it is the key data structure in pandas.it allows us to store and manipulate tabular data as a 2D data structures.


Different ways to create dataframe


1.   Through reading csv file
2.   Through reading excel file

1.   Through python dictionary
2.   Through python tuple list

1.  Through python dictionary








Create a simple Pandas DataFrame:

In [ ]:
import pandas as pd
data={'calories':[389,420,440],
      'duration':[50,40,45]}
df=pd.DataFrame(data)
df

,calories,duration
0,389,50
1,420,40
2,440,45


###Create dataframe using Python dictionary

In [ ]:
import pandas as pd
emp={'id':[101,102,103,104,105],
     'name':['deepak','shyam','arun','manu','jeena'],
     'age':[25,22,24,20,27],
     'salary':[25000,30000,150000,10000,300000]}
df=pd.DataFrame(emp)
df

,id,name,age,salary
0,101,deepak,25,25000
1,102,shyam,22,30000
2,103,arun,24,150000
3,104,manu,20,10000
4,105,jeena,27,300000


###Create datafrme using python tuple list

In [ ]:
import pandas as pd
empty_data=[(101,'deepak',25,25000),
            (102,'shyam',22,30000),
            (103,'arun',24,150000),
            (104,'manu',20,10000),
            (105,'jeena',27,30000)]
df=pd.DataFrame(empty_data,columns=['id','name','age','salary'])
df

,id,name,age,salary
0,101,deepak,25,25000
1,102,shyam,22,30000
2,103,arun,24,150000
3,104,manu,20,10000
4,105,jeena,27,30000


###Create Dataframe using a list of dictionaries

In [ ]:
import pandas as pd
emp_data_list=[{'id':101,'name':'deepak','age':25,'salary':25000},
               {'id':102,'name':'shyam','age':22,'salary':30000},
               {'id':103,'name':'arun','age':24,'salary':150000},
               {'id':104,'name':'manu','age':20,'salary':10000},
               {'id':105,'name':'jeena','age':27,'salary':30000}]
df=pd.DataFrame(emp_data_list)
df

,id,name,age,salary
0,101,deepak,25,25000
1,102,shyam,22,30000
2,103,arun,24,150000
3,104,manu,20,10000
4,105,jeena,27,30000


##Pandas Series

A pandas series like a column in a table

it is one-dimensional array holding data of any type.

Create a simple Pandas Series from a list:

In [ ]:
import pandas as pd

a=[1,7,2]
myvar=pd.Series(a)
print(myvar)

0    1
1    7
2    2
dtype: int64


##Labels

If nothing  else is specified,the values are labeled with their index number.First value has index 0,second value has index 1 etc.

In [ ]:
import pandas as pd
a=[1,7,2]
myvar=pd.Series(a)
print(myvar)
print(myvar[0])

0    1
1    7
2    2
dtype: int64
1


In [ ]:
import pandas as pd
a=[1,7,2]
myvar=pd.Series(a,index=['x','y','z'])
print(myvar)
print(myvar['y'])

x    1
y    7
z    2
dtype: int64
7


###Key Value Objects as series


In [ ]:
import pandas as pd
calories={'day1':420,'day2':300,'day3':390}
myvar=pd.Series(calories)
print(myvar)

day1    420
day2    300
day3    390
dtype: int64


In [ ]:
import pandas as pd
calories={'day1':420,'day2':300,'day3':390}
myvar=pd.Series(calories,index=['day1','day2'])
print(myvar)

day1    420
day2    300
dtype: int64


##Read txt file


In [ ]:
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/data.csv')
print(df)


     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


##Read excel file

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/employee_sample.xlsx')
print(df)

   ID     NAME DEPARTMENT
0   1    Alice         HR
1   2      Bob    Finance
2   3  Charlie         IT


##Skip the unnecessary header or add yours

In [ ]:
import pandas as pd
data=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/employee_salary_data.xlsx',skiprows=1)
print(data)

   101    Alice         HR  50000
0  102      Bob    Finance  60000
1  103  Charlie         IT  70000
2  104    David  Marketing  55000
3  105      Eva         IT  72000


**Pandas - Cleaning Data**

Data set:-

In [ ]:
import pandas as pd
data=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
print(data)

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

->**Empty Cells**

Empty cells can potentially give you a wrong result when you analyze data.



Remove Rows:-

The result from the converting in the example above gave us a NaT value, which can be handled as a NULL value, and we can remove the row by using the dropna() method.


eg :-Return a new Data Frame with no empty cell

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
new_df=df.dropna()
print(new_df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
19        60

 the dropna() method returns a new DataFrame, and will not change the original.

If you want to change the original DataFrame, use the inplace = True argument:

eg :-Remove all rows with NULL values:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
df.dropna(inplace=True)
print(df.to_string)

<bound method DataFrame.to_string of     Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17   

the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containing NULL values from the original DataFrame.

**Replace Empty Values**

Another way of dealing with empty cells is to insert a new value instead.


This way you do not have to delete entire rows just because of some empty cells.


The fillna() method allows us to replace empty cells with a value:

 eg:-Replace NULL values with the number 130:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
df.fillna(130,inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

**Replace Only For Specified Columns**

The example above replaces all empty cells in the whole Data Frame.


To only replace empty values for one column, specify the column name for the DataFrame:

eg:-Replace NULL values in the "Calories" columns with the number 130:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
df.fillna({'Calories':130},inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

**Replace Using Mean, Median, or Mode**

A common way to replace empty cells, is to calculate the mean, median or mode value of the column.


Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

EG:-Calculate the MEAN, and replace any empty values with it:

eg:-Calculate the MEAN, and replace any empty values with it:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
x=df['Calories'].mean()
df.fillna({'Calories':x},inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130    409.10
1         60  2020/12/02    117       145    479.00
2         60  2020/12/03    103       135    340.00
3         45  2020/12/04    109       175    282.40
4         45  2020/12/05    117       148    406.00
5         60  2020/12/06    102       127    300.00
6         60  2020/12/07    110       136    374.00
7        450  2020/12/08    104       134    253.30
8         30  2020/12/09    109       133    195.10
9         60  2020/12/10     98       124    269.00
10        60  2020/12/11    103       147    329.30
11        60  2020/12/12    100       120    250.70
12        60  2020/12/12    100       120    250.70
13        60  2020/12/13    106       128    345.30
14        60  2020/12/14    104       132    379.30
15        60  2020/12/15     98       123    275.00
16        60  2020/12/16     98       120    215.20
17        60  2020/12/17    100       120    300.00
18        45

eg:-Calculate the MEDIAN, and replace any empty values with it:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
x=df['Calories'].median()
df.fillna({'Calories':x},inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

eg:-Calculate the MODE, and replace any empty values with it:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
x=df['Calories'].mode()[0]
df.fillna({'Calories':x},inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

**Convert Into a Correct Format**

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_sample.xlsx')
print(df)

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

Check out row 22 and 26, the 'Date' column should be a string that represents a date:

to convert all cells in the 'Date' column into dates.

Pandas has a to_datetime() method for this:

eg;-Convert to date:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_data.xlsx')
df['Date']=pd.to_datetime(df['Date'],format='mixed')
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

the empty date in row 22 got a NaT (Not a Time) value, in other words an empty value. One way to deal with empty values is simply removing the entire row.

 so,remove rows with a NULL value in the "Date" column:

In [ ]:
df.dropna(subset=['Date'],inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7        450  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

**Pandas - Fixing Wrong Data**

---



**Wrong Data**

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

In [ ]:
import pandas as pd
df=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/workout_sample.xlsx')
df.loc[7,'Duration']=45
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7         45  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
18        45

In [ ]:
df.dropna(inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7         45  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
12        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
19        60

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

eg:-Loop through all values in the "Duration" column.

If the value is higher than 120, set it to 120:

In [ ]:
for x in df.index:
  if df.loc[x,'Duration']>120:
    df.loc[x,'Duration']=120

Another way of handling wrong data is to remove the rows that contains wrong data.

eg :Delete rows where "Duration" is higher than 120

In [ ]:
for x in df.index:
  if df.loc[x,"Duration"]>120:
    df.drop(x,inplace=True)

**Pandas - Removing
Duplicates**

---


**Discovering Duplicates**

Duplicate rows are rows that have been registered more than one time.



By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

 eg:Returns True for every row that is a duplicate, otherwise False:

In [ ]:
print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
29    False
30    False
31    False
dtype: bool


To remove duplicates, use the drop_duplicates() method.

In [ ]:
df.drop_duplicates(inplace=True)
print(df.to_string())

    Duration        Date  Pulse  Maxpulse  Calories
0         60  2020/12/01    110       130     409.1
1         60  2020/12/02    117       145     479.0
2         60  2020/12/03    103       135     340.0
3         45  2020/12/04    109       175     282.4
4         45  2020/12/05    117       148     406.0
5         60  2020/12/06    102       127     300.0
6         60  2020/12/07    110       136     374.0
7         45  2020/12/08    104       134     253.3
8         30  2020/12/09    109       133     195.1
9         60  2020/12/10     98       124     269.0
10        60  2020/12/11    103       147     329.3
11        60  2020/12/12    100       120     250.7
13        60  2020/12/13    106       128     345.3
14        60  2020/12/14    104       132     379.3
15        60  2020/12/15     98       123     275.0
16        60  2020/12/16     98       120     215.2
17        60  2020/12/17    100       120     300.0
19        60  2020/12/19    103       123     323.0
20        45

The (inplace = True) will make sure that the method does NOT return a new DataFrame, but it will remove all duplicates from the original DataFrame.

##Using Converters

Using converters to convert data using conditions



In [ ]:
#convert occupation of doctor to teacher
def convert(occupation):
  if occupation=='doctor':
    return 'teacher'
  else:
    return occupation

data=pd.read_excel('/content/drive/MyDrive/Colab Notebooks/dataset/sample_data_doctor.xlsx',converters={'occupation':convert})
print(data)

    name                age occupation
0   john                 23   Engineer
1  alice                 30    teacher
2   ammu  not available age    teacher


##Write to excel

use to_excel()

In [ ]:
data.to_excel('occupation.xlsx')

In [ ]:
data

,name,age,occupation
0,john,23,Engineer
1,alice,30,teacher
2,ammu,not available age,teacher


###Write two data frames in a single excel file



to write two data frames in a single excel file use ExcelWriter() method

In [ ]:
import pandas as pd

# Example DataFrames
data1=pd.DataFrame({'Name':['john','alice'],
                    'age':[23,30]})
data2=pd.DataFrame({'product':['pen','notebook'],
                    'price':[10,50]})

with pd.ExcelWriter('new file 1.xlsx')as wr:
  data1.to_excel(wr,sheet_name='sheet1')
  data2.to_excel(wr,sheet_name='sheet2')


In [ ]:
data1

,Name,age
0,john,23
1,alice,30


In [ ]:
data2

,product,price
0,pen,10
1,notebook,50


In [ ]:
import pandas as pd
data=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/airlines_flights_data.csv')
data

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
0,0,SpiceJet,SG-8709,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,1,SpiceJet,SG-8157,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,2,AirAsia,I5-764,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956
3,3,Vistara,UK-995,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,4,Vistara,UK-963,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955
...,...,...,...,...,...,...,...,...,...,...,...,...
300148,300148,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,69265
300149,300149,Vistara,UK-826,Chennai,Afternoon,one,Night,Hyderabad,Business,10.42,49,77105
300150,300150,Vistara,UK-832,Chennai,Early_Morning,one,Night,Hyderabad,Business,13.83,49,79099
300151,300151,Vistara,UK-828,Chennai,Early_Morning,one,Evening,Hyderabad,Business,10.00,49,81585


In [ ]:
data.head()

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
0,0,SpiceJet,SG-8709,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,1,SpiceJet,SG-8157,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,2,AirAsia,I5-764,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956
3,3,Vistara,UK-995,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,4,Vistara,UK-963,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955


In [ ]:
data.head(3)

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
0,0,SpiceJet,SG-8709,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,1,SpiceJet,SG-8157,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,2,AirAsia,I5-764,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956


In [ ]:
data.tail()

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
300148,300148,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,69265
300149,300149,Vistara,UK-826,Chennai,Afternoon,one,Night,Hyderabad,Business,10.42,49,77105
300150,300150,Vistara,UK-832,Chennai,Early_Morning,one,Night,Hyderabad,Business,13.83,49,79099
300151,300151,Vistara,UK-828,Chennai,Early_Morning,one,Evening,Hyderabad,Business,10.00,49,81585
300152,300152,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,81585


In [ ]:
data.tail(3)

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
300150,300150,Vistara,UK-832,Chennai,Early_Morning,one,Night,Hyderabad,Business,13.83,49,79099
300151,300151,Vistara,UK-828,Chennai,Early_Morning,one,Evening,Hyderabad,Business,10.00,49,81585
300152,300152,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,81585


In [ ]:
data.shape

(300153, 12)

In [ ]:
data.info()#to print details about the data,to find null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300153 entries, 0 to 300152
Data columns (total 12 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   index             300153 non-null  int64  
 1   airline           300153 non-null  object 
 2   flight            300153 non-null  object 
 3   source_city       300153 non-null  object 
 4   departure_time    300153 non-null  object 
 5   stops             300153 non-null  object 
 6   arrival_time      300153 non-null  object 
 7   destination_city  300153 non-null  object 
 8   class             300153 non-null  object 
 9   duration          300153 non-null  float64
 10  days_left         300153 non-null  int64  
 11  price             300153 non-null  int64  
dtypes: float64(1), int64(3), object(8)
memory usage: 27.5+ MB


In [ ]:
data.describe() #print all statistical results


,index,duration,days_left,price
count,300153.000000,300153.000000,300153.000000,300153.000000
mean,150076.000000,12.221021,26.004751,20889.660523
std,86646.852011,7.191997,13.561004,22697.767366
min,0.000000,0.830000,1.000000,1105.000000
25%,75038.000000,6.830000,15.000000,4783.000000
50%,150076.000000,11.250000,26.000000,7425.000000
75%,225114.000000,16.170000,38.000000,42521.000000
max,300152.000000,49.830000,49.000000,123071.000000


In [ ]:
data[['airline','price']]  #to selsect multiple column( 2 dimension table)

,airline,price
0,SpiceJet,5953
1,SpiceJet,5953
2,AirAsia,5956
3,Vistara,5955
4,Vistara,5955
...,...,...
300148,Vistara,69265
300149,Vistara,77105
300150,Vistara,79099
300151,Vistara,81585


In [ ]:
data[5:20]

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
5,5,Vistara,UK-945,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.33,1,5955
6,6,Vistara,UK-927,Delhi,Morning,zero,Morning,Mumbai,Economy,2.08,1,6060
7,7,Vistara,UK-951,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.17,1,6060
8,8,GO_FIRST,G8-334,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.17,1,5954
9,9,GO_FIRST,G8-336,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.25,1,5954
10,10,GO_FIRST,G8-392,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.25,1,5954
11,11,GO_FIRST,G8-338,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.33,1,5954
12,12,Indigo,6E-5001,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.17,1,5955
13,13,Indigo,6E-6202,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.17,1,5955
14,14,Indigo,6E-549,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.25,1,5955


In [ ]:
data[5:100:4]

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
5,5,Vistara,UK-945,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.33,1,5955
9,9,GO_FIRST,G8-336,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.25,1,5954
13,13,Indigo,6E-6202,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.17,1,5955
17,17,Air_India,AI-665,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.17,1,5955
21,21,GO_FIRST,G8-101,Delhi,Early_Morning,one,Night,Mumbai,Economy,14.50,1,5954
25,25,Vistara,UK-933,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.17,1,6690
29,29,Indigo,6E-5041,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,10470
33,33,Vistara,UK-817,Delhi,Evening,one,Morning,Mumbai,Economy,16.17,1,12150
37,37,Air_India,AI-453,Delhi,Early_Morning,one,Afternoon,Mumbai,Economy,8.83,1,11900
41,41,Air_India,AI-502,Delhi,Afternoon,one,Morning,Mumbai,Economy,19.08,1,12150


In [ ]:
data['price'].mean()

np.float64(20889.660523133203)

In [ ]:
data[(data['price']>10000) & (data['price']<20000)]

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
28,28,SpiceJet,SG-8169,Delhi,Evening,zero,Night,Mumbai,Economy,2.33,1,10260
29,29,Indigo,6E-5041,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,10470
30,30,GO_FIRST,G8-165,Delhi,Early_Morning,one,Afternoon,Mumbai,Economy,8.00,1,10575
31,31,Indigo,6E-2373,Delhi,Afternoon,one,Evening,Mumbai,Economy,6.00,1,10838
32,32,Vistara,UK-813,Delhi,Evening,one,Morning,Mumbai,Economy,14.67,1,12150
...,...,...,...,...,...,...,...,...,...,...,...,...
299945,299945,Air_India,AI-545,Chennai,Evening,zero,Night,Hyderabad,Business,1.42,45,15000
299987,299987,Air_India,AI-545,Chennai,Evening,zero,Night,Hyderabad,Business,1.42,46,15000
300029,300029,Air_India,AI-545,Chennai,Evening,zero,Night,Hyderabad,Business,1.42,47,15000
300069,300069,Air_India,AI-545,Chennai,Night,zero,Night,Hyderabad,Business,1.33,48,15000


In [ ]:
data.set_index(data['index'])

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
index,,,,,,,,,,,,
0,0,SpiceJet,SG-8709,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,1,SpiceJet,SG-8157,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,2,AirAsia,I5-764,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956
3,3,Vistara,UK-995,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,4,Vistara,UK-963,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955
...,...,...,...,...,...,...,...,...,...,...,...,...
300148,300148,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,69265
300149,300149,Vistara,UK-826,Chennai,Afternoon,one,Night,Hyderabad,Business,10.42,49,77105
300150,300150,Vistara,UK-832,Chennai,Early_Morning,one,Night,Hyderabad,Business,13.83,49,79099


**Loading data set**

In [ ]:
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/airlines_flights_data.csv')
df

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
0,0,SpiceJet,SG-8709,Delhi,Evening,zero,Night,Mumbai,Economy,2.17,1,5953
1,1,SpiceJet,SG-8157,Delhi,Early_Morning,zero,Morning,Mumbai,Economy,2.33,1,5953
2,2,AirAsia,I5-764,Delhi,Early_Morning,zero,Early_Morning,Mumbai,Economy,2.17,1,5956
3,3,Vistara,UK-995,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,4,Vistara,UK-963,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955
...,...,...,...,...,...,...,...,...,...,...,...,...
300148,300148,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,69265
300149,300149,Vistara,UK-826,Chennai,Afternoon,one,Night,Hyderabad,Business,10.42,49,77105
300150,300150,Vistara,UK-832,Chennai,Early_Morning,one,Night,Hyderabad,Business,13.83,49,79099
300151,300151,Vistara,UK-828,Chennai,Early_Morning,one,Evening,Hyderabad,Business,10.00,49,81585


##Group by

to get list of group by airlines

In [ ]:
group_result=df.groupby('airline')
list(group_result)

[('AirAsia',
           index  airline   flight source_city departure_time stops  \
  2            2  AirAsia   I5-764       Delhi  Early_Morning  zero   
  18          18  AirAsia   I5-747       Delhi        Evening   one   
  19          19  AirAsia   I5-747       Delhi        Evening   one   
  27          27  AirAsia   I5-744       Delhi        Morning   one   
  48          48  AirAsia   I5-784       Delhi        Evening   one   
  ...        ...      ...      ...         ...            ...   ...   
  206538  206538  AirAsia  I5-1229     Chennai          Night   one   
  206539  206539  AirAsia  I5-2462     Chennai          Night   one   
  206591  206591  AirAsia   I5-517     Chennai        Morning  zero   
  206616  206616  AirAsia  I5-1229     Chennai          Night   one   
  206617  206617  AirAsia  I5-2462     Chennai          Night   one   
  
           arrival_time destination_city    class  duration  days_left  price  
  2       Early_Morning           Mumbai  Economy   

to get airline with specific name


In [ ]:
group_result.get_group('Vistara')

,index,airline,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
3,3,Vistara,UK-995,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.25,1,5955
4,4,Vistara,UK-963,Delhi,Morning,zero,Morning,Mumbai,Economy,2.33,1,5955
5,5,Vistara,UK-945,Delhi,Morning,zero,Afternoon,Mumbai,Economy,2.33,1,5955
6,6,Vistara,UK-927,Delhi,Morning,zero,Morning,Mumbai,Economy,2.08,1,6060
7,7,Vistara,UK-951,Delhi,Afternoon,zero,Evening,Mumbai,Economy,2.17,1,6060
...,...,...,...,...,...,...,...,...,...,...,...,...
300148,300148,Vistara,UK-822,Chennai,Morning,one,Evening,Hyderabad,Business,10.08,49,69265
300149,300149,Vistara,UK-826,Chennai,Afternoon,one,Night,Hyderabad,Business,10.42,49,77105
300150,300150,Vistara,UK-832,Chennai,Early_Morning,one,Night,Hyderabad,Business,13.83,49,79099
300151,300151,Vistara,UK-828,Chennai,Early_Morning,one,Evening,Hyderabad,Business,10.00,49,81585


using forloop gruoping by airlines,flight,source city

In [ ]:
for flight,source_city in group_result:
  print(flight)
  print(source_city)

AirAsia
         index  airline   flight source_city departure_time stops  \
2            2  AirAsia   I5-764       Delhi  Early_Morning  zero   
18          18  AirAsia   I5-747       Delhi        Evening   one   
19          19  AirAsia   I5-747       Delhi        Evening   one   
27          27  AirAsia   I5-744       Delhi        Morning   one   
48          48  AirAsia   I5-784       Delhi        Evening   one   
...        ...      ...      ...         ...            ...   ...   
206538  206538  AirAsia  I5-1229     Chennai          Night   one   
206539  206539  AirAsia  I5-2462     Chennai          Night   one   
206591  206591  AirAsia   I5-517     Chennai        Morning  zero   
206616  206616  AirAsia  I5-1229     Chennai          Night   one   
206617  206617  AirAsia  I5-2462     Chennai          Night   one   

         arrival_time destination_city    class  duration  days_left  price  
2       Early_Morning           Mumbai  Economy      2.17          1   5956  
18     

###Group operations

to get maximum values of airlines respepectively

In [ ]:
group_result.max()

,index,flight,source_city,departure_time,stops,arrival_time,destination_city,class,duration,days_left,price
airline,,,,,,,,,,,
AirAsia,206617,I5-996,Mumbai,Night,zero,Night,Mumbai,Economy,19.58,49,31917
Air_India,300147,AI-9991,Mumbai,Night,zero,Night,Mumbai,Economy,49.83,49,90970
GO_FIRST,206589,G8-962,Mumbai,Night,zero,Night,Mumbai,Economy,22.50,49,32803
Indigo,206650,6E-998,Mumbai,Night,zero,Night,Mumbai,Economy,15.42,49,31952
SpiceJet,206184,SG-9974,Mumbai,Night,zero,Night,Mumbai,Economy,27.92,49,34158
Vistara,300152,UK-996,Mumbai,Night,zero,Night,Mumbai,Economy,47.08,49,123071


separatively:-

In [ ]:
group_result.get_group('Vistara').max()

,0
index,300152
airline,Vistara
flight,UK-996
source_city,Mumbai
departure_time,Night
stops,zero
arrival_time,Night
destination_city,Mumbai
class,Economy
duration,47.08


###Group by multiple columns

group by source_city,destination_city

In [ ]:
grp=df.groupby(['flight','source_city'])
list(grp)

[(('0.00E+00', 'Bangalore'),
           index airline    flight source_city departure_time stops  \
  97026    97026  Indigo  0.00E+00   Bangalore        Evening   one   
  97174    97174  Indigo  0.00E+00   Bangalore        Evening   one   
  97222    97222  Indigo  0.00E+00   Bangalore        Evening   one   
  97247    97247  Indigo  0.00E+00   Bangalore        Evening   one   
  97410    97410  Indigo  0.00E+00   Bangalore        Evening   one   
  97470    97470  Indigo  0.00E+00   Bangalore        Evening   one   
  97603    97603  Indigo  0.00E+00   Bangalore        Evening   one   
  97604    97604  Indigo  0.00E+00   Bangalore        Evening   one   
  97793    97793  Indigo  0.00E+00   Bangalore        Evening   one   
  97798    97798  Indigo  0.00E+00   Bangalore        Evening   one   
  97979    97979  Indigo  0.00E+00   Bangalore        Evening   one   
  98095    98095  Indigo  0.00E+00   Bangalore        Evening   one   
  98174    98174  Indigo  0.00E+00   Bangalore  

##Concat
It is used to concat two or more dataframes

In [ ]:
import pandas as pd
tcs={'id':[101,102,103,104],'name':['mithun','dipin','jose','rahul'],
     'age':[25,26,27,29]}

wipro={'id':[101,102,103,104],
       'name':['dev','sreenath','sreerag','shafeel'],
       'age':[24,24,27,26]}

tcs_emp=pd.DataFrame(tcs)
wipro_emp=pd.DataFrame(wipro)
tcs_emp

,id,name,age
0,101,mithun,25
1,102,dipin,26
2,103,jose,27
3,104,rahul,29


In [ ]:
wipro_emp

,id,name,age
0,101,dev,24
1,102,sreenath,24
2,103,sreerag,27
3,104,shafeel,26


In [ ]:
pd.concat([tcs_emp,wipro_emp])

,id,name,age
0,101,mithun,25
1,102,dipin,26
2,103,jose,27
3,104,rahul,29
0,101,dev,24
1,102,sreenath,24
2,103,sreerag,27
3,104,shafeel,26


ignore index-To get common index after concatenation

In [ ]:
pd.concat([tcs_emp,wipro_emp],ignore_index=True)

,id,name,age
0,101,mithun,25
1,102,dipin,26
2,103,jose,27
3,104,rahul,29
4,101,dev,24
5,102,sreenath,24
6,103,sreerag,27
7,104,shafeel,26


keys-To provide seperate keys for each concatenate data frames

In [ ]:
data=pd.concat([tcs_emp,wipro_emp],keys=['tcs','wipro'])
data

id      name  age
tcs   0  101    mithun   25
      1  102     dipin   26
      2  103      jose   27
      3  104     rahul   29
wipro 0  101       dev   24
      1  102  sreenath   24
      2  103   sreerag   27
      3  104   shafeel   26

concatenating vertically

In [ ]:
data=pd.concat([tcs_emp,wipro_emp],axis=1)
data

,id,name,age,id,name,age
0,101,mithun,25,101,dev,24
1,102,dipin,26,102,sreenath,24
2,103,jose,27,103,sreerag,27
3,104,rahul,29,104,shafeel,26


concatinating a dataframe with a series

In [ ]:
salary=pd.Series([65000,75000,15000,25000],name='salary')
pd.concat([tcs_emp,salary],axis=1)

,id,name,age,salary
0,101,mithun,25,65000
1,102,dipin,26,75000
2,103,jose,27,15000
3,104,rahul,29,25000


##Merge
Merge is a process to merge two tables horizontally

In [ ]:
import pandas as pd
tcs={'id':[101,102,103,106],'name':['mithun','dipin','joe','rahul'],
     'age':[25,26,27,29]}

wipro={'id':[101,102,103,105],
       'name':['dev','sreenath','sreerag','shafeel'],
       'age':[24,24,27,26]}

tcs_emp=pd.DataFrame(tcs)
wipro_emp=pd.DataFrame(wipro)

tcs_emp=pd.DataFrame(tcs)
tcs_emp

,id,name,age
0,101,mithun,25
1,102,dipin,26
2,103,joe,27
3,106,rahul,29


In [ ]:
wipro_emp=pd.DataFrame(wipro)
wipro_emp


,id,name,age
0,101,dev,24
1,102,sreenath,24
2,103,sreerag,27
3,105,shafeel,26


In [ ]:
pd.merge(tcs_emp,wipro_emp,on='id')

,id,name_x,age_x,name_y,age_y
0,101,mithun,25,dev,24
1,102,dipin,26,sreenath,24
2,103,joe,27,sreerag,27


only  respective number will appear if on='id' given

suffixes

In [ ]:
pd.merge(tcs_emp,wipro_emp,on='id', suffixes=['left','right'])

,id,nameleft,ageleft,nameright,ageright
0,101,mithun,25,dev,24
1,102,dipin,26,sreenath,24
2,103,joe,27,sreerag,27


outer join

In [ ]:
pd.merge(tcs_emp,wipro_emp,on='id',how='outer')

,id,name_x,age_x,name_y,age_y
0,101,mithun,25.0,dev,24.0
1,102,dipin,26.0,sreenath,24.0
2,103,joe,27.0,sreerag,27.0
3,105,NaN,NaN,shafeel,26.0
4,106,rahul,29.0,NaN,NaN


In [ ]:
pd.merge(tcs_emp,wipro_emp,on='id',how='outer',indicator=True)

,id,name_x,age_x,name_y,age_y,_merge
0,101,mithun,25.0,dev,24.0,both
1,102,dipin,26.0,sreenath,24.0,both
2,103,joe,27.0,sreerag,27.0,both
3,105,NaN,NaN,shafeel,26.0,right_only
4,106,rahul,29.0,NaN,NaN,left_only


left join

In [ ]:
pd.merge(tcs_emp,wipro_emp,on='id',how='left')

,id,name_x,age_x,name_y,age_y
0,101,mithun,25,dev,24.0
1,102,dipin,26,sreenath,24.0
2,103,joe,27,sreerag,27.0
3,106,rahul,29,NaN,NaN


right join

In [ ]:
pd.merge(tcs_emp,wipro_emp,on='id',how='right')

,id,name_x,age_x,name_y,age_y
0,101,mithun,25.0,dev,24
1,102,dipin,26.0,sreenath,24
2,103,joe,27.0,sreerag,27
3,105,NaN,NaN,shafeel,26


##Pivot and Pivot tables
Pivot allows to reshape the data frame

pivot table is used to summarize or aggregate data with in a dataframe

In [ ]:
import pandas as pd
datas=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/weather_dataset.csv')
datas

,Date,City,Temperature,Humidity,Wind Speed (km/h),Weather Condition
0,2025-08-01,Mumbai,30,85,12,Rainy
1,2025-08-01,Delhi,36,40,18,Sunny
2,2025-08-01,Bengaluru,27,70,10,Cloudy
3,2025-08-02,Mumbai,29,88,14,Rainy
4,2025-08-02,Delhi,35,45,20,Sunny
5,2025-08-02,Bengaluru,26,72,9,Cloudy
6,2025-08-03,Mumbai,28,90,11,Rainy
7,2025-08-03,Delhi,34,50,22,Partly Cloudy
8,2025-08-03,Bengaluru,27,68,8,Cloudy


create a table with date as index and city as columns

In [ ]:
datas.pivot(index='Date',columns='City')

Temperature               Humidity              Wind Speed (km/h)  \
City         Bengaluru Delhi Mumbai Bengaluru Delhi Mumbai         Bengaluru   
Date                                                                           
2025-08-01          27    36     30        70    40     85                10   
2025-08-02          26    35     29        72    45     88                 9   
2025-08-03          27    34     28        68    50     90                 8   

                        Weather Condition                        
City       Delhi Mumbai         Bengaluru          Delhi Mumbai  
Date                                                             
2025-08-01    18     12            Cloudy          Sunny  Rainy  
2025-08-02    20     14            Cloudy          Sunny  Rainy  
2025-08-03    22     11            Cloudy  Partly Cloudy  Rainy

create a table with Temperature

In [ ]:
datas.pivot(index='Date',columns='City',values='Temperature')

City,Bengaluru,Delhi,Mumbai
Date,,,
2025-08-01,27,36,30
2025-08-02,26,35,29
2025-08-03,27,34,28


used to summarize or aggregate data with in a dataframetas.pivot_table(index='City',columns='Date' ,aggfunc='sum')

In [ ]:
datas.pivot_table(index='City',columns='Date',aggfunc='sum',margins=True)

Humidity                            Temperature             \
Date      2025-08-01 2025-08-02 2025-08-03  All  2025-08-01 2025-08-02   
City                                                                     
Bengaluru         70         72         68  210          27         26   
Delhi             40         45         50  135          36         35   
Mumbai            85         88         90  263          30         29   
All              195        205        208  608          93         90   

                          Weather Condition                    \
Date      2025-08-03  All        2025-08-01        2025-08-02   
City                                                            
Bengaluru         27   80            Cloudy            Cloudy   
Delhi             34  105             Sunny             Sunny   
Mumbai            28   87             Rainy             Rainy   
All               89  272  RainySunnyCloudy  RainySunnyCloudy   

                                     \
Date                     2025-08-03   
City                                  
Bengaluru                    Cloudy   
Delhi                 Partly Cloudy   
Mumbai                        Rainy   
All        RainyPartly CloudyCloudy   

                                                              \
Date                                                     All   
City                                                           
Bengaluru                                 CloudyCloudyCloudy   
Delhi                                SunnySunnyPartly Cloudy   
Mumbai                                       RainyRainyRainy   
All        RainySunnyCloudyRainySunnyCloudyRainyPartly Cl...   

          Wind Speed (km/h)                             
Date             2025-08-01 2025-08-02 2025-08-03  All  
City                                                    
Bengaluru                10          9          8   27  
Delhi                    18         20         22   60  
Mumbai                   12         14         11   37  
All                      40         43         41  124

###Pivot Grouper

In [ ]:
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/Sales_data.csv')

df

,OrderID,Date,Category,Product,Quantity,Price
0,1001,2023-01-15,Electronics,Laptop,2,70000
1,1002,2023-01-20,Furniture,Chair,4,1500
2,1003,2023-02-05,Electronics,Headphones,5,2000
3,1004,2023-02-18,Furniture,Table,1,5000
4,1005,2023-03-03,Stationery,Notebook,10,50
5,1006,2023-03-25,Electronics,Mouse,6,800
6,1007,2023-04-10,Furniture,Sofa,1,15000
7,1008,2023-04-15,Stationery,Pen,20,20
8,1009,2023-05-01,Electronics,Laptop,1,70000
9,1010,2023-05-12,Furniture,Chair,2,1500


add total sales column

In [ ]:
df['Total']=df['Quantity']*df['Price']
df

,OrderID,Date,Category,Product,Quantity,Price,Total
0,1001,2023-01-15,Electronics,Laptop,2,70000,140000
1,1002,2023-01-20,Furniture,Chair,4,1500,6000
2,1003,2023-02-05,Electronics,Headphones,5,2000,10000
3,1004,2023-02-18,Furniture,Table,1,5000,5000
4,1005,2023-03-03,Stationery,Notebook,10,50,500
5,1006,2023-03-25,Electronics,Mouse,6,800,4800
6,1007,2023-04-10,Furniture,Sofa,1,15000,15000
7,1008,2023-04-15,Stationery,Pen,20,20,400
8,1009,2023-05-01,Electronics,Laptop,1,70000,70000
9,1010,2023-05-12,Furniture,Chair,2,1500,3000


create pivot table with grouper(monthly)

In [ ]:
df['Date'] = pd.to_datetime(df['Date'])
pivot =pd.pivot_table(df,
    index=pd.Grouper(key='Date',freq='ME'),
    columns='Category',
    values='Total',
    aggfunc='sum')
pivot

Category,Electronics,Furniture,Stationery
Date,,,
2023-01-31,140000.0,6000.0,NaN
2023-02-28,10000.0,5000.0,NaN
2023-03-31,4800.0,NaN,500.0
2023-04-30,NaN,15000.0,400.0
2023-05-31,70000.0,3000.0,NaN


##Melt

Melt allows to transform or reshape the data frame.

In [ ]:
import pandas as pd
datas=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/weather_dataset.csv')
datas

,Date,City,Temperature,Humidity,Wind Speed (km/h),Weather Condition
0,2025-08-01,Mumbai,30,85,12,Rainy
1,2025-08-01,Delhi,36,40,18,Sunny
2,2025-08-01,Bengaluru,27,70,10,Cloudy
3,2025-08-02,Mumbai,29,88,14,Rainy
4,2025-08-02,Delhi,35,45,20,Sunny
5,2025-08-02,Bengaluru,26,72,9,Cloudy
6,2025-08-03,Mumbai,28,90,11,Rainy
7,2025-08-03,Delhi,34,50,22,Partly Cloudy
8,2025-08-03,Bengaluru,27,68,8,Cloudy


In [ ]:
pd.melt(datas,id_vars=['Date'])

,Date,variable,value
0,2025-08-01,City,Mumbai
1,2025-08-01,City,Delhi
2,2025-08-01,City,Bengaluru
3,2025-08-02,City,Mumbai
4,2025-08-02,City,Delhi
5,2025-08-02,City,Bengaluru
6,2025-08-03,City,Mumbai
7,2025-08-03,City,Delhi
8,2025-08-03,City,Bengaluru
9,2025-08-01,Temperature,30


In [ ]:
melted=pd.melt(datas,
               id_vars=['City','Date','Weather Condition'],
               value_vars=['Temperature','Humidity'],
               var_name='Weather_Parameter',
               value_name='Reading')
melted


,City,Date,Weather Condition,Weather_Parameter,Reading
0,Mumbai,2025-08-01,Rainy,Temperature,30
1,Delhi,2025-08-01,Sunny,Temperature,36
2,Bengaluru,2025-08-01,Cloudy,Temperature,27
3,Mumbai,2025-08-02,Rainy,Temperature,29
4,Delhi,2025-08-02,Sunny,Temperature,35
5,Bengaluru,2025-08-02,Cloudy,Temperature,26
6,Mumbai,2025-08-03,Rainy,Temperature,28
7,Delhi,2025-08-03,Partly Cloudy,Temperature,34
8,Bengaluru,2025-08-03,Cloudy,Temperature,27
9,Mumbai,2025-08-01,Rainy,Humidity,85


##Cross tab or contingency table

In pandas, pd.crosstab() is used to compute a cross-tabulation of two or more factors. It’s often used for analyzing categorical variables, similar to a contingency table in statistics.

syntax:

pd.crosstab(index, columns, values=None, aggfunc=None, margins=False, margins_name='All')


In [ ]:
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/survey_crosstab.csv')
df

,RespondentID,Gender,AgeGroup,Preferred_Device
0,1,Male,18-25,Laptop
1,2,Female,26-35,Mobile
2,3,Male,18-25,Tablet
3,4,Female,18-25,Mobile
4,5,Male,26-35,Laptop
5,6,Female,18-25,Mobile
6,7,Male,36-45,Laptop
7,8,Female,26-35,Tablet
8,9,Male,26-35,Mobile
9,10,Female,36-45,Laptop


create cross_tabulation

In [ ]:
ct=pd.crosstab(df['Gender'],df['Preferred_Device'])
print(ct)

Preferred_Device  Laptop  Mobile  Tablet
Gender                                  
Female                 1       3       1
Male                   3       1       1


Cross-tab of AgeGroup vs Preferred_Device with totals:

In [ ]:
ct2=pd.crosstab(df['AgeGroup'],df['Preferred_Device'],margins=True,margins_name='Total')
print(ct2)

Preferred_Device  Laptop  Mobile  Tablet  Total
AgeGroup                                       
18-25                  1       2       1      4
26-35                  1       2       1      4
36-45                  2       0       0      2
Total                  4       4       2     10


##to_datetime()

to_datetime function is very useful in pandas to convert any string date to type datetime

In [ ]:
import pandas as pd
pd.to_datetime(['05/01/2017'],dayfirst=True)

DatetimeIndex(['2017-01-05'], dtype='datetime64[ns]', freq=None)

In [ ]:
dates=['2023-01-01','invalid_date','2023-03-30']
pd.to_datetime(dates,errors='coerce')

DatetimeIndex(['2023-01-01', 'NaT', '2023-03-30'], dtype='datetime64[ns]', freq=None)

In [ ]:
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/dataset/weather_dataset.csv')
df

,Date,City,Temperature,Humidity,Wind Speed (km/h),Weather Condition
0,2025-08-01,Mumbai,30,85,12,Rainy
1,2025-08-01,Delhi,36,40,18,Sunny
2,2025-08-01,Bengaluru,27,70,10,Cloudy
3,2025-08-02,Mumbai,29,88,14,Rainy
4,2025-08-02,Delhi,35,45,20,Sunny
5,2025-08-02,Bengaluru,26,72,9,Cloudy
6,2025-08-03,Mumbai,28,90,11,Rainy
7,2025-08-03,Delhi,34,50,22,Partly Cloudy
8,2025-08-03,Bengaluru,27,68,8,Cloudy


**Applymap**

 Use applymap() to increase all numeric values by 1.0 (if applicable)

We must exclude the 'City' column (non-numeric), or handle non-numeric safely

In [ ]:
print('original Dataframe:')
print(df)
df_numeric=df.select_dtypes(include='number')
df[df_numeric.columns]=df_numeric.applymap(lambda x:  x + 1)
print('\ndataframe after applymap(numeric values+1):')
df

original Dataframe:
         Date       City  Temperature  Humidity  Wind Speed (km/h)  \
0  2025-08-01     Mumbai           33        88                 15   
1  2025-08-01      Delhi           39        43                 21   
2  2025-08-01  Bengaluru           30        73                 13   
3  2025-08-02     Mumbai           32        91                 17   
4  2025-08-02      Delhi           38        48                 23   
5  2025-08-02  Bengaluru           29        75                 12   
6  2025-08-03     Mumbai           31        93                 14   
7  2025-08-03      Delhi           37        53                 25   
8  2025-08-03  Bengaluru           30        71                 11   

  Weather Condition  
0             Rainy  
1             Sunny  
2            Cloudy  
3             Rainy  
4             Sunny  
5            Cloudy  
6             Rainy  
7     Partly Cloudy  
8            Cloudy  

dataframe after applymap(numeric values+1):


/tmp/ipython-input-3594783373.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df[df_numeric.columns]=df_numeric.applymap(lambda x:  x + 1)


,Date,City,Temperature,Humidity,Wind Speed (km/h),Weather Condition
0,2025-08-01,Mumbai,34,89,16,Rainy
1,2025-08-01,Delhi,40,44,22,Sunny
2,2025-08-01,Bengaluru,31,74,14,Cloudy
3,2025-08-02,Mumbai,33,92,18,Rainy
4,2025-08-02,Delhi,39,49,24,Sunny
5,2025-08-02,Bengaluru,30,76,13,Cloudy
6,2025-08-03,Mumbai,32,94,15,Rainy
7,2025-08-03,Delhi,38,54,26,Partly Cloudy
8,2025-08-03,Bengaluru,31,72,12,Cloudy
